File: tests/_integration/smessage_encryption.py
#
# Copyright (c) 2017 Cossack Labs Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import sys
from base64 import b64encode, b64decode
from pythemis import smessage

_, COMMAND, SENDER_PRIVATE_KEY, RECIPIENT_PUBLIC_KEY, MESSAGE = range(5)

if len(sys.argv) != 5:
    print('Usage: <command: enc | dec | sign | verify > <send_private_key> <recipient_public_key> <message>')
    exit(1)

command = sys.argv[COMMAND]
private_key_path = sys.argv[SENDER_PRIVATE_KEY]
public_key_path = sys.argv[RECIPIENT_PUBLIC_KEY]
message = sys.argv[MESSAGE]

with open(private_key_path, 'rb') as f:
    private_key = f.read()

with open(public_key_path, 'rb') as f:
    public_key = f.read()

message_encrypter = smessage.SMessage(private_key, public_key)

if command == 'enc':
    encrypted = message_encrypter.wrap(message.encode('utf-8'))
    encoded = b64encode(encrypted)
    print(encoded.decode('ascii'))
elif command == 'dec':
    decoded = b64decode(message.encode('utf-8'))
    decrypted = message_encrypter.unwrap(decoded)
    print(decrypted.decode('utf-8'))
elif command == 'sign':
    encrypted = smessage.ssign(private_key, message.encode('utf-8'))
    encoded = b64encode(encrypted)
    print(encoded.decode('ascii'))
elif command == 'verify':
    decoded = b64decode(message.encode('utf-8'))
    decrypted = smessage.sverify(public_key, decoded)
    print(decrypted.decode('utf-8'))
else:
    print('Wrong command, use <enc | dev | sign | verify>')
    exit(1)

File: tests/_integration/smessage_encryption.go
package main

import (
	"encoding/base64"
	"fmt"
	smessage "github.com/cossacklabs/themis/gothemis/message"
	"os"
	"io/ioutil"
	"github.com/cossacklabs/themis/gothemis/keys"
)

func main() {
	args := os.Args

	if len(args) != 5 {
		fmt.Printf("Usage: %s <command: enc | dec | sign | verify > <send_private_key> <recipient_public_key> <message>\n", os.Args[0])
		os.Exit(1)
	}

	command := args[1]
	private_key_path := args[2]
	private_key, err := ioutil.ReadFile(string(private_key_path))
	if err != nil {
		panic(err)
	}
	public_key_path := args[3]
	public_key, err := ioutil.ReadFile(string(public_key_path))
	if err != nil {
		panic(err)
	}
	message := args[4]

	message_encrypter := smessage.New(&keys.PrivateKey{Value: private_key}, &keys.PublicKey{Value: public_key})

	if "enc" == command {
		encData, err := message_encrypter.Wrap([]byte(message))
		if nil != err {
			fmt.Println(os.Stderr, "Error encrypting message")
			os.Exit(1)
		}
		fmt.Println(base64.StdEncoding.EncodeToString(encData))

	} else if "dec" == command {
		decoded_message, err := base64.StdEncoding.DecodeString(message)
		if nil != err {
			fmt.Println(os.Stderr, "Error decoding message")
			os.Exit(1)
		}

		decData, err := message_encrypter.Unwrap(decoded_message)
		if nil != err {
			fmt.Println(os.Stderr, "Error decrypting message")
			os.Exit(1)
		}
		fmt.Println(string(decData[:]))

	} else if "sign" == command {
        encData, err := message_encrypter.Sign([]byte(message))
        if nil != err {
            fmt.Println(os.Stderr, "Error encrypting message")
            os.Exit(1)
        }
        fmt.Println(base64.StdEncoding.EncodeToString(encData))
    } else if "verify" == command {
        decoded_message, err := base64.StdEncoding.DecodeString(message)
        if nil != err {
            fmt.Println(os.Stderr, "Error decoding message")
            os.Exit(1)
        }

        decData, err := message_encrypter.Verify(decoded_message)
        if nil != err {
            fmt.Println(os.Stderr, "Error decrypting message")
            os.Exit(1)
        }
        fmt.Println(string(decData[:]))

    } else {
		fmt.Println(os.Stderr, "Wrong command, use <enc | dev | sign | verify>")
		os.Exit(1)
	}
}
